src/frequenz/sdk/actor/data_sourcing/microgrid_source.py

```diff
@@ -0,0 +1,373 @@
+
+        Returns:
+            A dictionary of output metric names to channel senders from the
+            channel registry.
+        """
+        return [
+            (
+                self._get_data_extraction_method(category, metric),
+                [
+                    self._registry.get_sender(request.get_channel_name())
+                    for request in reqlist
+                ],
+            )
+            for (metric, reqlist) in requests.items()
+        ]
+
+    async def _handle_data_stream(
+        self,
+        comp_id: int,
+        category: ComponentCategory,
+    ) -> None:
+        """Stream component data and send the requested metrics out.
+
+        Args:
+            comp_id: Id of the requested component.
+            category: The category of the component.
+        """
+        stream_senders = []
+        if comp_id in self._req_streaming_metrics:
+            await self._check_requested_component_and_metrics(
+                comp_id, category, self._req_streaming_metrics[comp_id]
+            )
+            stream_senders = self._get_metric_senders(
+                category, self._req_streaming_metrics[comp_id]
+            )
+        api_data_receiver = self.comp_data_receivers[comp_id]
+
+        def process_msg(data: Any) -> None:
+            tasks = []
+            for (extractor, senders) in stream_senders:
+                for sender in senders:
+                    tasks.append(sender.send(Sample(data.timestamp, extractor(data))))
+            asyncio.gather(*tasks)
+
+        async for data in api_data_receiver:
+            process_msg(data)
+
+    async def _update_streams(
+        self,
+        comp_id: int,
+        category: ComponentCategory,
+    ) -> None:
+        """Update the requested metric streams for the given component.
+
+        Args:
+            comp_id: Id of the requested component.
+            category: Category of the requested component.
+        """
+        if comp_id in self.comp_data_tasks:
+            self.comp_data_tasks[comp_id].cancel()
+
+        self.comp_data_tasks[comp_id] = asyncio.create_task(
+            self._handle_data_stream(comp_id, category)
+        )
+
+    async def add_metric(
+        self, category: ComponentCategory, request: ComponentMetricRequest
+    ) -> None:
+        """Add a metric to be streamed from the microgrid API.
+
+        Args:
+            category: Category of the requested component.
+            request: A request object for a metric, received from a downstream
+                actor.
+        """
+        comp_id = request.component_id
+
+        if comp_id not in self._req_streaming_metrics:
+            self._req_streaming_metrics[comp_id] = {request.metric: [request]}
+        elif request.metric not in self._req_streaming_metrics[comp_id]:
+            self._req_streaming_metrics[comp_id][request.metric] = [request]
+        else:
+            self._req_streaming_metrics[comp_id][request.metric].append(request)
```

Comment:
pro-tip: You can use [`dict.setdefault()`](_BatteryDataMethods) for this.

```suggestion
        self._req_streaming_metrics.setdefault(comp_id, {}).setdefault(request.metric, []).append(request)
```

Not sure if it makes it more clear or more opaque, though :)

Reply:
it is more readable,  and people can look up the method,  updated, thanks. :-)